### PR TITLE
RavenDB-6097 wrapping Tcp connections with SslStream

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: src/Raven.Client/Documents/TcpUtils.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using Raven.Client.Server.Commands;
+
+namespace Raven.Client.Document
+{
+    public static class TcpUtils
```
Comment (ppekrol) [inline src/Raven.Client/Documents/TcpUtils.cs:12]: Can you make this internal?

File: src/Raven.Client/Documents/TcpUtils.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        public static async Task<Stream> WrapStreamWithSsl(TcpClient tcpClient, TcpConnectionInfo info)
+        {
+            Stream stream = tcpClient.GetStream();
+            if (info.Certificate != null)
+            {
+                var Cert = new X509Certificate2(Convert.FromBase64String(info.Certificate));
```
Comment (ayende) [inline src/Raven.Client/Documents/TcpUtils.cs:19]: Why this casing?

File: src/Raven.Client/Documents/TcpUtils.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                var sslStream = new SslStream(stream, false, (sender, certificate, chain, errors) => Cert.Equals(certificate));
+                await
+                    sslStream.AuthenticateAsClientAsync("RavenDB", new X509Certificate2Collection { Cert }, SslProtocols.Tls12,
```
Comment (ayende) [inline src/Raven.Client/Documents/TcpUtils.cs]: You _don't_ authenticate from the client side

File: src/Raven.Client/Documents/TcpUtils.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Client/Documents/TcpUtils.cs:12]: YEs

File: src/Raven.Server/CertificateUtils.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+            X509Certificate2 x509 = new X509Certificate2(certificate.GetEncoded());
+            x509.FriendlyName = subjectName;
+
+            CaPrivateKey = issuerKeyPair.Private;
+
+            return x509;
+        }
+
+        public static AsymmetricAlgorithm ToDotNetKey(RsaPrivateCrtKeyParameters privateKey)
```
Comment (ayende) [inline src/Raven.Server/CertificateUtils.cs]: Unused, should be removed

File: src/Raven.Server/CertificateUtils.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    {
+
+        public static X509Certificate2 CreateSelfSignedCertificate(string subjectName, string issuerName)
+        {
+            AsymmetricKeyParameter CaPrivateKey;
+            CreateCertificateAuthorityCertificate(subjectName, out CaPrivateKey);
+            return CreateSelfSignedCertificateBasedOnCertificateAuthorityPrivateKey(subjectName, issuerName, CaPrivateKey);
+        }
+
+        public static X509Certificate2 CreateSelfSignedCertificateBasedOnCertificateAuthorityPrivateKey(string subjectName, string issuerName, AsymmetricKeyParameter issuerPrivKey)
```
Comment (ayende) [inline src/Raven.Server/CertificateUtils.cs]: Should be private

File: src/Raven.Server/CertificateUtils.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            var convertedCertificate =
+                new X509Certificate2(
+                    stream.ToArray(), null,
+                    X509KeyStorageFlags.PersistKeySet | X509KeyStorageFlags.Exportable);
+            stream.Position = 0;
+
+            return convertedCertificate;
+        }
+
+        public static X509Certificate2 CreateCertificateAuthorityCertificate(string subjectName, out AsymmetricKeyParameter CaPrivateKey)
```
Comment (ayende) [inline src/Raven.Server/CertificateUtils.cs]: Should be private

File: src/Raven.Server/RavenServer.cs [hunk 7]
```diff
@@ -198,6 +201,30 @@ public void Initialize()
...
+            public X509Certificate2 Certificate;
+        }
+
+        private static CertificateHolder GenerateSelfSignedCertificate()
+        {
+            //TODO: remove this when https://github.com/dotnet/coreclr/issues/8148 is fixed
+            var @var = Environment.GetEnvironmentVariable("COMPlus_ReadyToRunExcludeList");
+            if (@var != "System.Security.Cryptography.X509Certificates")
+            {
+                throw new MissingMemberException("Missing environment variable COMPlus_ReadyToRunExcludeList setting, can't use SslStream on dotnet core 1.1.0");
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: Why missing member?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/RavenServer.cs (hunk 7 above)
Comment (talweiss1982) [reply src/Raven.Server/RavenServer.cs]: didn't know what to throw

**Comment by talweiss1982** [comment]
fixed

**Pull request merged**